tests/storage/test_state.py

```diff
@@ -483,3 +482,119 @@ def test_get_state_for_event(self):
 
         self.assertEqual(is_all, True)
         self.assertDictEqual({(e5.type, e5.state_key): e5.event_id}, state_dict)
+
+    def test_state_filter_difference(self):
+        def assert_difference(
+            minuend: StateFilter, subtrahend: StateFilter, expected: StateFilter
+        ):
+            self.assertEqual(
+                minuend.approx_difference(subtrahend),
+                expected,
+                f"StateFilter difference not correct:\n\n\t{minuend!r}\nminus\n\t{subtrahend!r}\nwas\n\t{minuend.approx_difference(subtrahend)}\nexpected\n\t{expected}",
+            )
+
+        # it's not possible to subtract individual state keys from
+        # a wildcard
+        assert_difference(
+            StateFilter.all(),
+            StateFilter.from_types(
+                ((EventTypes.Member, "@wombat:hs1"), (EventTypes.Member, "@spqr:hs1"))
+            ),
+            StateFilter.all(),
+        )
+        self.assertEqual(
+            StateFilter.all().approx_difference(
+                StateFilter.from_types(
+                    (
+                        (EventTypes.Member, "@wombat:hs1"),
+                        (EventTypes.Member, "@spqr:hs1"),
+                    )
+                )
+            ),
+            StateFilter.all(),
+        )
```

Comment:
I think this is repeating the same assertion twice in different ways?

Reply:
huh, you're right. I wonder why that got left in.